Open source rewrite of rqt_topic

flynneva opened the pull request:
- open sourced from Apex.AI

Closes #46

evan-flynn-apexai committed:
Open source rewrite of rqt_topic

- open sourced from Apex.AI

Signed-off-by: Evan Flynn <evan.flynn@apex.ai>

flynneva commented:
fyi @methylDragon @wjwwood for some reason the tests are failing after migrating to rolling, so give me a few days to fix that and then we should be good to get this reviewed / merged in.

wjwwood commented:
I'll have to defer to @methylDragon on this one, but feel free to ping me if you need my opinion on anything @methylDragon.

flynneva commented:
@methylDragon @clalancette feel free to have a review of this PR. It's just a start on the rewrite for sure, but I think it gets this GUI in a much more usable state than before 👍🏼 

Interested to hear your thoughts!

Also - happy to help maintain this repo if this PR is accepted.

flynneva commented on test/models/test_message_detail.py:
```diff
@@ -0,0 +1,141 @@
+
+
+def test_message_detail_model(message_detail, qtmodeltester):
+    # qtmodeltester.check(message_detail)
```
I'll make a follow-up issue to tackle re-enabling these checks if this PR is merged in. I'm not quite sure why its failing all of a sudden once we open sourced this tool and dont have the time right now to dig into it.

clalancette commented:
> @methylDragon @clalancette feel free to have a review of this PR. It's just a start on the rewrite for sure, but I think it gets this GUI in a much more usable state than before 👍🏼

I have a few high-level comments here:
1.  This is great!  rqt_topic definitely needs some love.
2.  There's a lot of changing of `'` to `"` in this PR, which doesn't conform to our current style.  This means this will almost certainly fail our flake8 tests, which need to pass for this to get in.
3.  If we are going to heavily refactor this package, I think we should also refactor it to use more of our best practices.  In particular, the layout of the code should conform much more closely to our recommended best practice.  You can see that in any of the packages in e.g. https://github.com/ros2/ros2cli .
4.  Finally, I'm not a huge fan of complete rewrites.  They are difficult to review and hard to explain to users.  I'd much rather see this as a series of improvements to the current code, with the goal of getting towards something like this PR.  That will make it much more likely for us to get this in.  If I were going to do this, then what I'd do is to first refactor the code as I mentioned in 3, then do a series of PRs where you introduce/refactor one part at a time.

Let me know if this makes sense to you.

flynneva commented:
> There's a lot of changing of ' to " in this PR, which doesn't conform to our current style. This means this will almost certainly fail our flake8 tests, which need to pass for this to get in.

@clalancette sure, this is an easy search and replace 👍🏼 

> If we are going to heavily refactor this package, I think we should also refactor it to use more of our best practices. In particular, the layout of the code should conform much more closely to our recommended best practice. You can see that in any of the packages in e.g. https://github.com/ros2/ros2cli .

@clalancette what "layout" in particular are you referring to in the linked repo? The linked repo contains _many_ packages...but this repo only contains one. Are you suggesting to split up `rqt_topic` into many smaller packages? Or do you just mean removing the `src` directory and having `rqt_topic` at the top level?

> Finally, I'm not a huge fan of complete rewrites. They are difficult to review and hard to explain to users. I'd much rather see this as a series of improvements to the current code, with the goal of getting towards something like this PR. That will make it much more likely for us to get this in. If I were going to do this, then what I'd do is to first refactor the code as I mentioned in 3, then do a series of PRs where you introduce/refactor one part at a time.

@clalancette sure, I agree as well. I just did it this way because internally the rewrite was already complete.  If you'd like me to break this up into smaller chunks I can...just it will take much longer probably (months) to get this complete.

clalancette commented:
> @clalancette what "layout" in particular are you referring to in the linked repo? The linked repo contains _many_ packages...but this repo only contains one. Are you suggesting to split up `rqt_topic` into many smaller packages? Or do you just mean removing the `src` directory and having `rqt_topic` at the top level?

The latter; removing `src` and having `rqt_topic` at the top level.

> @clalancette sure, I agree as well. I just did it this way because internally the rewrite was already complete. If you'd like me to break this up into smaller chunks I can...just it will take much longer probably (months) to get this complete.

I think it is going to take months either way.  But with smaller chunks we can actually see forward progress, which I think is a good thing.

flynneva commented:
@clalancette took care of the `src` dir thing and migrating to use single quotes `'`. How would you like this PR split up? There is no real good way to "slowly" migrate without breaking things. Just let me know how you'd like to get this split up and I can do it 👍🏼

evan-flynn-apexai committed:
Add new folder to hold common button QAction objects

Signed-off-by: Evan Flynn <evan.flynn@apex.ai>

evan-flynn-apexai committed:
Add background worker objects

- meant to be ran in background threads instead of main GUI thread

Signed-off-by: Evan Flynn <evan.flynn@apex.ai>

evan-flynn-apexai committed:
Add new rqt_topic view objects

- Views are split up into three main widgets
- Topic list view that lists all available topics and their info
- Message list view that lists messages from all selected topics
- Message detail view that shows a detailed tree view of a single
  message

Signed-off-by: Evan Flynn <evan.flynn@apex.ai>

evan-flynn-apexai committed:
Add new rqt_topic models, with tests

Signed-off-by: Evan Flynn <evan.flynn@apex.ai>

evan-flynn-apexai committed:
Bring rqt_topic rewrite together, use in main

Signed-off-by: Evan Flynn <evan.flynn@apex.ai>

evan-flynn-apexai commented:
> But with smaller chunks we can actually see forward progress,

@clalancette I've broken up the single big commit into smaller more manageable ones for you to review. Let me know if you'd like me to split these out into their own MRs, though I dont think we can do this without breaking things.

P.s. sorry about my github accounts - I always forget which one I'm signed into. :sweat:

ahcorde committed:
override subscriber QoS

Signed-off-by: Alejandro Hernandez Cordero <ahcorde@gmail.com>

ahcorde committed:
Merge remote-tracking branch 'ros/rolling' into rolling

ahcorde committed:
Added copyright

Signed-off-by: Alejandro Hernandez Cordero <ahcorde@gmail.com>

ahcorde committed:
Added contributing an licence

Signed-off-by: Alejandro Hernandez Cordero <ahcorde@gmail.com>

ahcorde committed:
Added missing dependency

Signed-off-by: Alejandro Hernandez Cordero <ahcorde@gmail.com>

ahcorde commented:
Pulls: ros-visualization/rqt_topic#47
Gist: https://gist.githubusercontent.com/ahcorde/1a4b43a2b562340cc0b3612871218aae/raw/0d183da8c07752aae0700a98305767dabcb8677a/ros2.repos
BUILD args:  --packages-above-and-dependencies rqt_topic
TEST args:  --packages-above rqt_topic
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16567
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24409)](http://ci.ros2.org/job/ci_linux/24409/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18481)](http://ci.ros2.org/job/ci_linux-aarch64/18481/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=3854)](http://ci.ros2.org/job/ci_linux-rhel/3854/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=24974)](http://ci.ros2.org/job/ci_windows/24974/)

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24414)](http://ci.ros2.org/job/ci_linux/24414/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18484)](http://ci.ros2.org/job/ci_linux-aarch64/18484/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=3857)](http://ci.ros2.org/job/ci_linux-rhel/3857/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=24977)](http://ci.ros2.org/job/ci_windows/24977/)

ahcorde commented:
Related PRs 
 - https://github.com/ros2/ros2/pull/1711
 - https://github.com/ros2/ci/pull/820

flynneva commented:
@ahcorde let me know if you need any help getting this merged! Glad to see some interest in getting this in

ahcorde committed:
Merge branch 'rolling' into rolling

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24600)](http://ci.ros2.org/job/ci_linux/24600/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18646)](http://ci.ros2.org/job/ci_linux-aarch64/18646/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=4057)](http://ci.ros2.org/job/ci_linux-rhel/4057/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25142)](http://ci.ros2.org/job/ci_windows/25142/)

ahcorde commented:
> @ahcorde let me know if you need any help getting this merged! Glad to see some interest in getting this in

Thank you for the contribution, I think I included the required changes in CI to be able to merge it!

It's looking super great!

christophebedard commented on test/test_flake8.py:
```diff
@@ -19,5 +19,6 @@
 def test_flake8():
-    rc = main(argv=[])
-    assert rc == 0, 'Found errors'
+    pass
+    # rc = main(argv=[])
+    # assert rc == 0, 'Found errors'
```
Changes to this file should probably be reverted

ahcorde replied:
PR https://github.com/ros-visualization/rqt_topic/pull/58
